### DISCUSSION TITLE
Memtest86+ ECC memory support

### DISCUSSION THREAD

**Discussion by UttarwarSunil** [memtest86plus/memtest86plus#248] [Q&A]
Hi Team,

I see that MemTest86 from Passmark support ECC memory and reports the errors as shown in below snapshot.
![ecc-memtest86](https://user-images.githubusercontent.com/70257439/215079254-f12974ff-d029-4d40-98f0-c3cc1836be64.png)

Do we have ECC memory support in Memtest86+? Can you please confirm MemTest86+ also reports similar ECC memory errors?

Thanks
Sunil

**Comment by debrouxl** [comment]
Nope, reporting ECC-corrected memory errors is not currently supported in memtest86+. See #92 , which you can upvote if EDAC support is really important to you :)

**Reply by UttarwarSunil** [reply]
Do we have any timeline or any release plan for this feature support? It will be very helpful.

**Reply by x86fr** [reply]
Soon for one CPU type, maybe long ago for your particular CPU.

I will start working on a single dev platform soon, to implement timings read and ECC/EDAC registers, but every different CPU/chipset requires a specific code path.

**Comment by x86fr** [comment]
Unrelated, but if anybody find a defective ECC module able to generate correctable ECC errors, i'm looking for one to implement this function.

**Reply by 01e3** [reply]
Hello there :) 

https://github.com/memtest86plus/memtest86plus/issues/63#issuecomment-1198980259

I just need to find our where did I put it...

**Reply by 01e3** [reply]
Found it! Proof ;)

![image](https://user-images.githubusercontent.com/105136494/215294561-3a9c44a4-3045-455d-b723-3e51056f9297.png)

**Reply by x86fr** [reply]
Do you think the motherboard can boot with only a single module?

**Reply by 01e3** [reply]
Yes, I can do it in my Lenovo S30 (C602, NCT6681, E5-2667 v2), no problem. 

The only issue is that SPD does not work there. :/

**Reply by 01e3** [reply]
I think I may also have a faulty 1GB PC2-4200E module somewhere - same story: generating correctable ECC errors.

Usable on my Dell PowerEdge 840 (Intel 3000, Xeon 3070)

However, I have been unable to locate it so far. :/

**Comment by debrouxl** [comment]
I've seen occasional corrected ECC errors with a 16 GB HMT42GR7AFR4A-PB stick in one of my RS904A servers. It's a matter of once in a couple days or so, and I have no trace of any such error in recent kernel logs, though: as such, the stick, if/when I find it, may be a poor fit for attempting to reproduce ECC errors. I'll see if I can obtain the corrected ECC error trace again, but it will probably take a while, as these machines seldom reach days of uptime, let alone at full load.

**Comment by UttarwarSunil** [comment]
We dont need actually erroneous ECC dimm, we can simulate the error. I observed that memtest86 from passmark is having inject option where it can simulate the errors.
![image](https://user-images.githubusercontent.com/70257439/217585913-1e22fc43-5427-4b09-bd96-eae9db11a3d2.png)

**Reply by x86fr** [reply]
ECC injection is highly CPU dependent, often disabled or not present on retail hardware (it's a server-class debugging features) and requires specific code to handle it (without saying that this code is commonly under NDA). Nowadays, ECC injection is considered a security issue, so manufacturers are strongly encouraged to lock it. 

At early stage of development, it's MUCH easier to generate real errors with a defective module than to use another code (which may be bugged) to generate false errors.

**Comment by UttarwarSunil** [comment]
Hi Team,

We are looking for ECC support in memtest86+ and want to expedite the feature implementation. 

Is there anything needed on this, please let us know. 

Can we have discussion on this over call/mail? Please share details want to discuss.

Thanks
Sunil

**Reply by x86fr** [reply]
I'm working on a first implementation that will target Skylake specifically. What's your target CPU?

**Reply by UttarwarSunil** [reply]
We are looking to have support for AMD ryzen CPU. Please let me know we can discuss over call/mail.

**Reply by oed-streicherd** [reply]
Can test on large scale hosts; we have some ECC DIMMs (DDR4).
Testing on Cisco Server >3TB Memory
Cascade Lake or Sky Lake multi socket (2/4) hosts

give me bump when testing is possible

**Comment by debrouxl** [comment]
Sam has started publishing code, so far for members of the Zen series: https://github.com/memtest86plus/memtest86plus/tree/x86fr/ecc .
